mm2src/mm2_p2p/src/behaviours/atomicdex.rs

```diff
@@ -50,6 +55,18 @@ const ANNOUNCE_INTERVAL: Duration = Duration::from_secs(600);
 const ANNOUNCE_INITIAL_DELAY: Duration = Duration::from_secs(60);
 const CHANNEL_BUF_SIZE: usize = 1024 * 8;
 
+/// Used in time validation logic for each peer which runs immediately  after the
+/// `ConnectionEstablished` event.
+const MAX_TIME_GAP_FOR_CONNECTED_PEER: u64 = 30;
```

Comment:
I am not sure if this is a good default. Should we make it like 20 or 25?

Reply:
30 seconds are fine since it's half the max allowed difference for swaps.. https://github.com/KomodoPlatform/komodo-defi-framework/blob/f487947c9d8f454969396986f8f259961ce87f6b/mm2src/mm2_main/src/lp_swap.rs#L154 
However, An edge case that can happen in theory is time differences accumulating across the network. Consider a scenario with peers A, B, C and D:

- Peers B and C are relay/seed nodes
- Peers A and D are starting a swap as taker/maker
- If time differences exist as follows:
  - Peer A is 30 secs off peer B but they allow them to connect
  - Peer D is also 30 secs off peer C but they allow them to connect
  - Peer B and C are out of sync as well (any number of seconds less than 30)
  - Peer A and D can have more than 60 seconds difference and the swap fails.

This is an edge case that we should keep in mind in case we got time difference errors in a swap after this fix, but I think it will happen in very rare cases if any.

Reply:
P.S. let's make it 20 secs to reduce the probability of this edge case happening.

Reply:
The edge case can only happen if seed nodes are not synced in time, which shouldn't be the case I believe? Since we are the maintainers of the seed network.

> P.S. let's make it 20 secs to reduce the probability of this edge case happening.

I can do that, but I am not sure if that's a good idea. I had ~9 seconds diff between my laptop and desktop computer, so I believe in some cases it could be even more than that.

Reply:
> The edge case can only happen if seed nodes are not synced in time, which shouldn't be the case I believe? Since we are the maintainers of the seed network.

Keep in mind that the 64 notary nodes are also seed nodes.

Reply:
> I can do that, but I am not sure if that's a good idea. I had ~9 seconds diff between my laptop and desktop computer, so I believe in some cases it could be even more than that.

Mmm, I see that you already made this 20 secs so let's leave it at that and see. Is there a way to inform the peer that their clock is off so that they can adjust it? Will look into that now as I am starting the full review.

Reply:
> Is there a way to inform the peer that their clock is off so that they can adjust it?

Just read internal discussions about this, the currently connected peers is one solution to this but not optimal, will think of something if there are any other solutions. A problem we have is the no login mode, GUIs should be able to show the orderbooks and other info for no login mode even if the peer clock is off, it will be strange for someone who wants to have a look at our dex to get a warning about system clock when not logged in.

Reply:
IMO it's totally reasonable to tell someone to sync their time if their system isn't (which shouldn't be the case for general-purpose OS users). There are so many apps doing this.

Reply:
> There are so many apps doing this.

Never seen it for web apps like komodo wallet though.

Reply:
I guess maybe you never run your OS unsynced in time :D

Reply:
> I guess maybe you never run your OS unsynced in time :D

You are probably right :D

> If we want to keep the network synced in time, I don't know what else we can do other than not connecting to unsynced ones.

I would leave the PR as is and if we got something from GUI team about this we can look for workarounds for this particular case. We need to mention it though while they are testing.